# More Fun with Functions

## The print Function

One of the nice things about Python is that there are a lot of built-in functions that have already been written for us. One built-in function we have already seen is "print". We've already seen that "print" can be used to show values of variable on the screen:

In [3]:
num = 5
print(num)

5


print can actually take arguments and show all of their values:

In [5]:
num1 = 5
num2 = 10
print(num1, num2, num1 + num2)

5 10 15


By default, print will put a space between each value, but you can change this by specifying the "sep" argument:

In [8]:
print(num1, num2, num3, sep='_')
print(num1, num2, num3, sep='/')
print(num1, num2, num3, sep='')
print(num1, num2, num3, sep=' ')

5_10_15
5/10/15
51015
5 10 15


### Sidenote: Optional Arguments and Default Values

The "sep" argument for the print function is an optional argument, and if you do not specify its value, it will default to a space. You can also create optional arguments with default values for your own functions like this:

In [12]:
def my_function(a, b=10, c=2):
    answer = a + b / c
    return answer

This function takes three arguments, $a$, $b$, and $c$, and it returns $a + b \div c$. $a$ is a required argument, but $b$ and $c$ are optional, with the default values of $5$ and $10$ respectively. You can call it just by giving $a$ as an argument:

In [13]:
print(my_function(0))

5.0


You can call it by just giving $a$ and $b$:

In [18]:
print(my_function(0, 16))

8.0


You can call it with $a$, $b$, and $c$: 

In [19]:
print(my_function(0, 16, 8))

2.0


Or if you just want to specify $c$ and leave $b$ with its default value, you can refer to $c$ by name:

In [21]:
print(my_function(0, c=5))

2.0


You always have to give it $a$, though:

In [23]:
print(my_function(c=5))

TypeError: my_function() missing 1 required positional argument: 'a'

## Other Built-in Functions

One of the nice things about Python is that it has many useful built-in functions besides just "print". "input" is another useful function that allows you to assign the value of the variable while the program is running:

In [1]:
num = input('Enter a number: ')
print('Your number is:', num)

Enter a number: 5
Your number is:  5


Some other useful functions include "abs" which returns the absolute value of a number:

In [2]:
num = -1
abs_num = abs(num)
print(abs_num)

1


"round" which takes two arguments. The first argument is the number you want to round, and the second is an optional argument that tells it how many decimals to round to (if you don't give this argument, it defaults to rounding to the nearest whole number).

In [4]:
num = 3.3333333333333
round1 = round(num, 2)
round2 = round(num)

print(round1)
print(round2)

3.33
3


and "min" and "max" which return the minimum and maximum of two numbers, respectively:

In [6]:
num1 = 100
num2 = 1000

mymin = min(num1, num2)
mymax = max(num1, num2)

print(mymin)
print(mymax)

100
1000


## Exercise

Write a Python script that takes two numbers from the user as input. Then it prints the maximum of the two numbers, rounded to three decimal places.

## Modules

Many more functions can be found inside of *modules*. Modules are collections of pre-written functions and variables. Python comes with many different kinds of modules. For example, let's look at the the *math* module. To load a module you need to use the "import" statement:

In [5]:
import math

To use functions and variables from this module, we need to use the "dot" (**.**) operator like this:

In [8]:
print(math.pi)

3.141592653589793


The "dot" operator tells Python to go inside of "math" and look for something called "pi". "pi" is one of the variables that can be found in the math module. Some other useful variables include "inf" for infinity, and "e" for Euler's number:

In [9]:
print(math.inf)
print(math.e)

inf
2.718281828459045


The math module has a lot of useful functions, too. It has functions to calculate square roots and logarithms:

In [11]:
print(math.sqrt(2)) #print the square root of 2
print(math.log10(1000)) #print the logarithm of 1000, base 10

1.4142135623730951
3.0


It also has trigonmetric functions like sin, and cos:

In [13]:
print(math.sin(math.pi))
print(math.cos(math.pi))

1.2246467991473532e-16
-1.0


To see everything the math module can do, you can read the documentation: https://docs.python.org/3/library/math.html

### Scientific Notation , Numerical Precision, and Binary

Notice that Python said $\sin(\pi) = 1.2246467991473532e-16$. You might be wondering what the "e-16" at the end means. This is Python's way of writing *scientific notation*. This is the same thing as $1.2246467991473532 \times 10 ^{-16}$, or $0.00000000000000012246467991473532$.

If you've taken Trigonometry, you might recall that $\sin(\pi) = 0$, so why isn't that the answer that Python gave you? The reason is because of *numerical precision*. $\pi$ is an irrational number, which means that it has an infinite number of digits. Python does not have infinite memory, though, so it can only remember a certain number of digits. That means that when it does calculations with $\pi$, there are going to be rounding errors.

Sometimes Python makes rounding errors even when there are only a few digits involved. For example, look what happens when you calculate $0.1 \times 3$. 

In [17]:
print(0.1 * 3)

0.30000000000000004


Why does Python behave this way? You might have heard before that computers represent everything as $1$'s and $0$'s. This is called *binary representation*. The number system that you are used to is called *decimal representation*. Here are some conversions between decimal and binary representations:

$0 = 0$

$1 = 1$

$2 = 10$

$3 = 11$

$4 = 100$

There is actually a built-in function in Python called "bin" that converts integers from decimal to binary. (The actual answer comes after the "0b"; the "0b" is just letting you know the number is in binary).

In [18]:
print(bin(2))
print(bin(256))
print(bin(12345))

0b10
0b100000000
0b11000000111001


Just like some numbers cannot be represented in a finite number of digits in decimal (like $\frac{1}{3}$) some numbers cannot be represented in a finite number of digits in binary (like $0.1$), and this causes rounding errors.

Make sure you keep these rounding errors in mind when you do calculations in Python! Fortunately, most of the time they're small enough to not matter that much. You can usually get the answer you want by making use of the "round" function:

In [21]:
print(round(math.sin(math.pi)))
print(round(0.1*3, 1))

0
0.3


## Exercise

Write a function that takes two arguments $x$ and $y$ and makes the following calculation:
$\sqrt x + 2\pi \times y$

Round your answer to 4 decimal places.

## Random Numbers

We can have Python create random numbers using the "random" module. It oftens several diffent functions for generating different kinds of random numbers:

In [3]:
import random

a = 1
b = 10

print(random.random()) #prints a random number between 0 and 1
print(random.uniform(a, b)) #prints a random number between a and b
print(random.randint(a, b)) #prints a random integer between a and b

0.025365946998282096
2.510673173036392
6


Try running the code a bunch of different times. You should get different results every time.

## Exercise

Write a function that takes two arguments $x$ and $y$ and makes the following calculation:
$\sqrt x + 2\pi \times y + z$

Where $z$ is a random integer between $x$ and $y$. Round your answer to 4 decimal places.

# More Fun with Variables

So far, we've been focusing on working with just numbers in Python, but Python can work with lots of different *data types*. For example, we can assign letters and words to variables:

In [34]:
a = 'a'
print(a)

dog = "dog"
print(dog)

a
dog


Groups of letters and words are called *strings* in Python. Strings always have to be surrounded by apostraphes (**'**) or quotation marks (**"**). We can add and multiply strings:

In [37]:
a = 'a'
b = 'b'

print(a + b)
print(b * 5)

ab
bbbbb


We can convert a number into a string using the built-in "str" function. Notice that adding strings together is different than adding numbers together:

In [40]:
one = str(1)
two_point_two = str(2.2)

print(one + two_point_two)

12.2


We can convert strings into whole-numbers using the "int" function, or into decimals with the "float" function:

In [42]:
one = int('1')
two_point_two = float('2.2')

print(one + two_point_two)

3.2


There are lots of cool things you can do with strings, but we'll talk more about the later. For now, just remember that strings are different from numbers, and you can assign strings to variables.

You can also assign functions to variables in Python. For example:

In [43]:
r = round
print(r(3.3))
print(r(1.2345, 2))

3
1.23


## Exercises

Try to figure out what the value of $x$ will be after running the following Python code. Then actually run the code to check your work:

In [6]:
x = 'x'
x = x * 5 + x

In [7]:
x = 3
x = 3 * str(x) + str(2 * x)

In [8]:
import math

y = round
z = math.pi

x = y(y(z) + z, 2)

## Getting Values as Input

What if we wanted to write some code that greets the user when they run the program? For example, if the user's name was "Luigi", the program would say "Hello, Luigi!", We can do this with the built-in variable "input":

In [5]:
name = input("Type in your name: ")
print("Hello, ", name, '!', sep='')

Type in your name: Luigi
Hello, Luigi!


"input" always returns a string, so if you want a number, you'll need to convert it to one:

In [78]:
#This way will treat the input as strings
number1 = input('Type in your first number: ')
number2 = input('Type in your second number: ')
print("The sum of your numbers is:", number1+number2)

Type in your first number: 1
Type in your second number: 2
The sum of your numbers is: 12


In [79]:
#This way will treat the input as numbers
number1 = input('Type in your first number: ')
number2 = input('Type in your second number: ')

number1 = float(number1)
number2 = float(number2)

print("The sum of your numbers is:", number1+number2)

Type in your first number: 1
Type in your second number: 2
The sum of your numbers is: 3.0


If you want to hide what the user is typing (like when you type in a password), you can use the "getpass" function from the "getpass" module:

In [4]:
import getpass

password = getpass.getpass("Type in your password: ")
print(password)

Type in your password: ········
hunter2


## Exercise

Write a Python script that asks for the user's name and favorite number as input, and then prints out "Hello, name, the squareroot of your number is $\sqrt x$  !", where x is the number given by the user.